essentials-module-teleport/src/main/kotlin/io/github/nyliummc/essentials/commands/RTPCommand.kt

```diff
@@ -56,17 +57,18 @@ object RTPCommand {
         val deltaZ = distance * player.random.nextGaussian() + 1
         val newX = deltaX + player.x
         val newZ = deltaZ + player.z
-        var newY = 0.0
         var i = world.seaLevel
 
-        while (!world.isAir(BlockPos(newX, i++.toDouble(), newZ)) && newY + world.seaLevel < world.height) {
-            newY++
-        }
-
         if (world.isWater(BlockPos(newX, (i - 2).toDouble(), newZ))) {
             return execute(context)
         }
 
+        var pos = BlockPos(newX, i++.toDouble(), newZ)
+        while ((!world.isAir(pos) || world.getBlockState(pos).block == Blocks.CAVE_AIR) && i < world.height) {
+            i++
+            pos = BlockPos(newX, i.toDouble(), newZ)
```

Comment:
newY never updated, just use the original loop.

Also, this loop should be before the water check just like before.